internal/gps/source_manager.go

```diff
@@ -455,6 +462,72 @@ func (sm *SourceMgr) DeduceProjectRoot(ip string) (ProjectRoot, error) {
 	return ProjectRoot(pd.root), err
 }
 
+// DeduceConstraint tries to puzzle out what kind of version is given in a string -
+// semver, a revision, or as a fallback, a plain tag
+func (sm *SourceMgr) DeduceConstraint(s string, pi ProjectIdentifier) (Constraint, error) {
+	if s == "" {
+		// Find the default branch
+		versions, err := sm.ListVersions(pi)
+		if err != nil {
+			return nil, errors.Wrapf(err, "list versions for %s(%s)", pi.ProjectRoot, pi.Source) // means repo does not exist
+		}
+
+		SortPairedForUpgrade(versions)
+		for _, v := range versions {
+			if v.Type() == IsBranch {
+				return v.Unpair(), nil
+			}
+		}
+	}
+
+	// always semver if we can
+	c, err := NewSemverConstraintIC(s)
+	if err == nil {
+		return c, nil
+	}
+
+	slen := len(s)
```

Comment:
By moving this into `*sourceMgr`, we create an opportunity to do better validation here. Rather than deducing whether it's a revision from what general rules we have about the patterns of the bytes, we can actually hand this off to a `sourceGateway`, which can validate that it's an appropriate revision _for that source type_.

However, that may end up being somewhat involved to really do well, so we can defer it to a later issue. (Let's open up a follow-up prior to merging this issue for doing that, though).